### ISSUE TITLE
Is it possible to use additional object attributes in model validation to obtain metrics based on them?

### ISSUE THREAD

**Issue by Altanus** [ultralytics/ultralytics#12126]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm currently using Ultralytics for key points detection. I've trained my model and I wonder if it is possible to get validation metrics separately for occluded objects and not occluded ones. 
And if it is possible to do the same but with custom objects attributes different from labels?

**Comment by glenn-jocher** [comment]
Hello! Great question! 👍

For obtaining validation metrics based on attributes like occlusion or other custom attributes, YOLOv8 directly does not separate metrics based on them during validation. However, you can implement a custom evaluation metric.

For instance, you could modify your validation set to include a flag for occlusion in your labels, and then during validation, segment the results based on these flags to compute metrics separately. Here’s a simple conceptual outline using Python:

```python
from ultralytics import YOLO, evaluate

# Load your model
model = YOLO('path/to/your/model.pt')

# Custom evaluation method to compute metrics based on occlusion
def evaluate_occlusion(data, occluded=False):
    results = model.val(data=data)
    if occluded:
        # Filter results for occluded objects based on your custom flag/attribute
        occluded_results = [r for r in results if r.attribute == 'occluded']
        metrics = evaluate(occluded_results)
    else:
        non_occluded_results = [r for r in results if r.attribute != 'occluded']
        metrics = evaluate(non_occluded_results)
    return metrics

# Usage
occluded_metrics = evaluate_occlusion('path/to/your/dataset.yaml', occluded=True)
non_occluded_metrics = evaluate_occlusion('path/to/your/dataset.yaml', occluded=False)
```

This example assumes your data can be filtered by an ‘attribute’ flag; adjust as per your data structure and requirements. This approach provides flexibility in handling and evaluating various object attributes. Happy coding! 😊

**Comment by Altanus** [comment]
Hello! Thank you for your answer. I think that's a great way to come up with the issue. And thank you for the code suggestion. 
Nevertheless I still have a few questions. 
1. Iterating through **results = model.val(data=data)** should give a list with metrics. Did you mean it? Or maybe I have to model.predict(data=data) to iterate over predictions?
2. You've mentioned **from ultralytics import YOLO, evaluate**. Did you mean some specific module? Can you provide me with an info where I can find it? Or it is expected me to code it by myself? (actually not the best choice as key point detection metrics are quite complicated)

**Comment by glenn-jocher** [comment]
Hello!

I'm delighted to clarify those points for you!

1. Apologies for any confusion. You should indeed use `model.predict(data=data)` to iteratively obtain predictions for each item in your dataset and then manually apply your evaluation logic based on occlusion or other criteria. `model.val()` is typically used to assess model performance utilizing the entire validation dataset and does not provide detailed, itemized data needed for attribute-specific metrics unless you manually extract and process these from the output.

2. Regarding the `evaluate` function, this was illustrative and assumes implementation of a custom function that handles detailed attribute-based performance metrics. If creating such a function sounds daunting, I recommend checking out the evaluation tools provided by popular libraries like NumPy or Scikit-Learn which can help in calculating metrics if you adjust the input to match the specific conditions (like occlusion).

Here’s a slight modification using `predict()`:
```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Custom evaluation method to compute metrics based on occlusion
def evaluate_occlusion(data, occluded=False):
    results = model.predict(data=data)  # Get predictions for each item
    occluded_results = [r for r in results if r.attribute == 'occluded']
    # Apply your evaluation logic to occluded_results here, such as accuracy, recall, etc.
    metrics = custom_evaluate(occluded_results)  # Implement your metrics calculation
    return metrics

def custom_evaluate(results):
    # Implement your evaluation logic here
    pass

# Usage
occluded_metrics = evaluate_occlusion('path/to/your/dataset.yaml', occluded=True)
```

I hope this helps streamline your development process! Let me know if there's anything else you need. 😊
